trie/hasher.go

```diff
@@ -50,8 +53,16 @@ func returnHasherToPool(h *hasher) {
 // original node initialzied with the computed hash to replace the original one.
 func (h *hasher) hash(n node, db DatabaseWriter, force bool) (node, node, error) {
 	// If we're not storing the node, just hashing, use avaialble cached data
-	if hash, dirty := n.cache(); hash != nil && (db == nil || !dirty) {
-		return hash, n, nil
+	if hash, dirty := n.cache(); hash != nil {
+		if db == nil {
+			return hash, n, nil
+		}
+		if n.canUnload(h.cachegen, h.cachelimit) {
```

Comment:
Could you add a comment somewhere explaining when cache eviction happens? If I'm understanding correctly, whenever a node is modified, we evaluate all its unmodified children for eviction, in effect. Meaning we could retain parts of the trie much longer than the cache policy implies, but as soon as a parent node is modified, we'll notice and discard any old nodes.

This also means that we could theoretically retain a node so long that the generation counter wraps around a 16 bit int and the node appears recent even though it's not. With 100 generation retention and a 16 bit int, this should happen at most 100/65536 = .15% of the time, which seems acceptable.

It's also an invariant, if I'm not mistaken, that the cache generation of a node is >= that of all its subnodes, meaning when a node expires, we can be certain all its subnodes are also expired and discardable; it'd be good to note that, too.